### PR TITLE
Feature/alg 287 refactor ssd

### PR DESCRIPTION
**Pull request by yurkovak**
Refactoring SSD to be based on CustomizableDetector. Many lines changed due to Black. Real changes are:
* Split SSD into backbone, neck and head, now all these modules can be replaced/reused with other backbones/necks/heads from the detection modules factory with minimum understanding of the required yaml structure;
* Moved customizable backbone, neck and head into a different location due to circular imports, other than migration this code didn't change;
* Add variable input_channels to `MobileNetV1`
* Fix inference of `MobilenetV2` in backbone mode
* Minor change in `MultiOutputModule` to support lists from yamls (`ListConfig` isn't an instance of `list`);
* Fix head replacement in customizable detector;
* In pretrained models test the only real change is this line:
```python
ssd_dboxes = DEFAULT_SSD_LITE_MOBILENET_V2_ARCH_PARAMS["heads"]["SSDHead"]["anchors"]
```
* updated pretrained weights in S3 (by the same path as before) to match the new state dict, ran the test

### CODE REVIEW

**Comment by yurkovak** [comment]
Black forced me to re-save these files, no real changes there:
```
would reformat src/super_gradients/examples/deci_platform_logger_example/deci_platform_logger_example.py
would reformat src/super_gradients/common/sg_loggers/deci_platform_sg_logger.py
would reformat src/super_gradients/sanity_check/env_sanity_check.py
would reformat src/super_gradients/training/utils/quantization_utils.py
```

File: src/super_gradients/training/models/classification_models/mobilenetv2.py [hunk 1]
```diff
@@ -151,17 +148,17 @@ def __init__(
...
+            return self.classifier(x)
 
     def _extract_connection_layers_input_channel_size(self):
         """
         Extracts the number of channels out when using mobilenetV2 as yolo backbone
         """
-        curr_layer_input = torch.rand(
-            1, self.in_channels, 320, 320
-        )  # input dims are used to extract number of channels
+        curr_layer_input = torch.rand(1, self.in_channels, 320, 320)  # input dims are used to extract number of channels
```
Comment (shaydeci) [inline src/super_gradients/training/models/classification_models/mobilenetv2.py:161]: Following my previous comment, do we still need this method?

File: src/super_gradients/training/models/classification_models/mobilenetv2.py [hunk 2]
```diff
@@ -128,20 +128,17 @@ def __init__(
...
+                    self.features.append(block(curr_channels, output_channel, 1, expand_ratio=t, grouped_conv_size=grouped_conv_size))
                 curr_channels = output_channel
         # building last several layers
         self.features.append(conv_1x1_bn(curr_channels, self.last_channel))
         # make it nn.Sequential
         self.features = nn.Sequential(*self.features)
+        self.backbone_mode = backbone_mode
 
-        if backbone_mode:
+        if self.backbone_mode:
```
Comment (shaydeci) [inline src/super_gradients/training/models/classification_models/mobilenetv2.py:141]: Why do we still need self.backbone_connection_channels attribute with the new design?

File: tests/integration_tests/pretrained_models_test.py [hunk 3]
```diff
@@ -220,613 +219,634 @@ def setUp(self) -> None:
...
+            valid_loader=self.transfer_classification_dataloader,
+        )
 
     def test_pretrained_ssd_lite_mobilenet_v2_coco(self):
-        trainer = Trainer('coco_ssd_lite_mobilenet_v2',
-                          multi_gpu=MultiGPUMode.OFF)
-        model = models.get("ssd_lite_mobilenet_v2",
-                           arch_params=self.coco_pretrained_arch_params["ssd_lite_mobilenet_v2"],
-                           **self.coco_pretrained_ckpt_params)
+        trainer = Trainer("coco_ssd_lite_mobilenet_v2", multi_gpu=MultiGPUMode.OFF)
```
Comment (shaydeci) [inline tests/integration_tests/pretrained_models_test.py:462]: Did you modify the checkpoint in our pretrained models bucket? Does the SSD tests here all pass?

File: src/super_gradients/modules/detection_modules.py [hunk 4]
```diff
@@ -20,10 +27,361 @@ def __init__(self, in_channels: Union[List[int], int], **kwargs):
     @property
     @abstractmethod
     def out_channels(self) -> Union[List[int], int]:
+        raise NotImplementedError()
+
+
+class NStageBackbone(BaseDetectionModule):
```
Comment (shaydeci) [inline src/super_gradients/modules/detection_modules.py:33]: Whats the use of this class in this PR..?

File: src/super_gradients/modules/detection_modules.py [hunk 5]
```diff
@@ -20,10 +27,361 @@ def __init__(self, in_channels: Union[List[int], int], **kwargs):
...
+        outputs = []
+        outputs_logits = []
+        for output, output_logits in preds:
+            outputs.append(output)
+            outputs_logits.append(output_logits)
+
+        return outputs if self.training else (torch.cat(outputs, 1), outputs_logits)
+
+
+class MobileNetV1Backbone(BaseDetectionModule):
```
Comment (shaydeci) [inline src/super_gradients/modules/detection_modules.py:173]: MobileNetV1Backbone and MobileNetV2Backbone are almost identical, can we perhaps introduce a parent class ?

File: src/super_gradients/training/models/classification_models/mobilenetv2.py [hunk 6]
```diff
@@ -151,17 +148,17 @@ def __init__(
 
     def forward(self, x):
         x = self.features(x)
-        x = x.mean(3).mean(2)
-        x = self.classifier(x)
-        return x
+        if self.backbone_mode:
+            return x
+        else:
+            x = x.mean(3).mean(2)
```
Comment (shaydeci) [inline src/super_gradients/training/models/classification_models/mobilenetv2.py:154]: Why did we remove the mean ops in backbone mode ? Wont this be problematic when one would manually replace head ?

**Review by shaydeci — CHANGES_REQUESTED** [review]
Some comments and clarifications inline.
Looks nice!

File: tests/integration_tests/pretrained_models_test.py (hunk 3 above)
Comment (yurkovak) [reply tests/integration_tests/pretrained_models_test.py:462]: Yes, renamed the old one instead. Tests pass. Or to be more specific, v2 does, while v1 didn't pass even before my changes. mAP of v1 is 0 because the pre-trained weights are for a dataset with [0, 1] normalization

File: src/super_gradients/modules/detection_modules.py (hunk 4 above)
Comment (yurkovak) [reply src/super_gradients/modules/detection_modules.py]: Gone, changed merged in #490

File: src/super_gradients/training/models/classification_models/mobilenetv2.py (hunk 2 above)
Comment (yurkovak) [reply src/super_gradients/training/models/classification_models/mobilenetv2.py:141]: In the new design, no, but I'm not sure about the old one, afraid to break something. Let's get back to this when we refactor yolos

File: src/super_gradients/training/models/classification_models/mobilenetv2.py (hunk 1 above)
Comment (yurkovak) [reply src/super_gradients/training/models/classification_models/mobilenetv2.py:161]: Will be happy to remove it later, looks bad

File: src/super_gradients/training/models/classification_models/mobilenetv2.py (hunk 6 above)
Comment (yurkovak) [reply src/super_gradients/training/models/classification_models/mobilenetv2.py:154]: I think that's the pattern we've been using everywhere in classification models: return a feature map in a backbone mode, apply avg pool -> FC otherwise. Why would you want to change a head of a model in a backbone mode?

File: src/super_gradients/modules/detection_modules.py (hunk 5 above)
Comment (yurkovak) [reply src/super_gradients/modules/detection_modules.py]: Agree, done e6172c5

File: src/super_gradients/training/models/classification_models/mobilenetv2.py (hunk 1 above)
Comment (shaydeci) [reply src/super_gradients/training/models/classification_models/mobilenetv2.py:161]: Please add a @TODO comment so wed remember to do so (hopefully).

File: src/super_gradients/training/models/classification_models/mobilenetv2.py (hunk 2 above)
Comment (shaydeci) [reply src/super_gradients/training/models/classification_models/mobilenetv2.py:141]: Please add a @todo comment so wed remember to do so (hopefully).

**Review by shaydeci — APPROVED** [review]
LGTM
